man/pconfig.1.rst
========
 pmerge
========

---------------------------------
pkgcore configuration inspection
---------------------------------

:Date:   2008-02-09
:Manual section: 1
:Manual group: pkgcore

SYNOPSIS
========

  pconfig command [options]

DESCRIPTION
===========

pconfig is the pkgcore configuration tool.  Dumping of the translated
configuration, listing the known configurables that can be used in
configuration, and general introspection of configurables is all via
this tool.

OPTIONS
=======

.. pkgcore_script_options:: pkgcore.scripts.pconfig.commandline_commands


NEWS
Pkgcore Release Notes
=====================

See ChangeLog for full commit logs; this is summarized/major changes.

* add tar contentsSet rewriting; tarballs sometimes leave out directories,
  and don't always have the fully resolved path- /usr/lib/blah, when
  /usr/lib -> /usr/lib64 *should* be /usr/lib64/blah, but tar doesn't force
  this.  Due to that, can lead to explosions in unpacking- this is now fixed.

* pquery --attr inherited was added; this feature may disappear down the
  line, adding it meanwhile since it's useful for ebuild devs.

* adjust setup.py so that man page installation properly respects --root

* correct a corner case where a package name of 'dev-3D' was flagged as
  invalid.


0.4:

* resolver fixes: vdb loadup wasn't occuring for old style virtuals for
  rdepend blockers, now forces it.  It was possible for a node to be
  considered usable before it's rdepends blockers were leveled- now those
  must be satisfied before being able to dep on the node.

* resolver events cleanup; pmerge now gives far better info as to why a
  choice failed, what it attempted to get around it, etc.

* multiplex trees now discern their frozen state from their subtrees,
  and will execute the repo_op for the leftmost subtree if unfrozen.

* pquery --attr eapi was added.

* ticket 94; package.provided is now supported fully both in profiles,
  and in user profile (/etc/portage/profile).

* ticket 116; ignore empty tarfile exception if the exception explicitly
  states empty header.

* utter corner case compatibility- =dev-util/diffball-1.0-r0 is now the
  same as =dev-util/diffball-1.0 .

* convert FETCHCOMMAND/RESUMECOMMAND support to execute spawn_bash by
  default instead of trying to cut out shell; this kills off the occasional
  incompatibility introduced via portage supplying make.globals.

* FEATURES=sfperms is now a trigger instead of a dyn_preinst hook.
  Faster, cleaner, etc.

* delayed unpacking of binpkgs has been disabled; occasionally can lead to
  quadratic behaviour in contents accessing, and extreme corner case trigger
  breakages.  Will be re-enabled once API has been refactored to remove
  these issues.

* FEATURES=multilib-strict was converted into a trigger.  Tries to
  use the python bindings for file first (merge file[python]), falling
  back to invoking file.  Strongly suggested you have the bindings- fair bit
  faster.  Finally, verification now runs for binpkgs also.

* bug 137; symlink on directory merging failures where pkgcore would wipe
  files it had just installed invalidly.

* correct issue in offset rewriting (was resetting new_offset to '/')-
  should only be api visible, no existing consumers known.

* ebuild env lzma unpack support was broken; fixed (ticket 140).

* Additional debug output for pmerge.

* Further extending PortageFormatter to sanely handle worldfile highlights
  and show repos with both id and location

* Ticket 132: Portage Formatter supports real portage colors now,
  thanks to agaffney for getting the ball rolling

* Masked IUSEs were not treated right in all cases, thanks to agaffney
  for report and help testing

* diefunc tracebacks beautified)


pkgcore 0.3.4:

* IUSEs were filtered, unstated were not respected though breaks with
  current portage tree, so re-enabling.
  Also sanely handle -flag enforcing now and kill hackish code for it.


pkgcore 0.3.3:

* IUSE defaults are respected now, so EAPI=1 implemented

* Write slotted atoms to worldfile as portage supports this now

* Sync up with portage; add support for lzma to unpack- mirror r7991 from
  portage.


pkgcore 0.3.2:

* ticket 190746 from gentoo; basically need to force the perms of first level
  directory of an unpacked $DISTDIR to ensure it's at least readable/writable.
  fixes unpacking of app-misc/screen-4.0.3_p20070403::gentoo-x86 .

* ticket 118; if -u, don't add the node to world set.

* correct a corner case in python implementation of cpv comparison (just
  python, cpy extension handles it correctly); bug 188449 in gentoo, basically
  floats have a limited precision, thus it was possible to get truncation in
  comparison with specially crafted versions.

* handle EOF/IOError on raw_input (for --ask) a bit more gracefully, ticket
  108.

* cd to ${WORKDIR} if ${S} doesn't exist for test/install phases; matches
  change in portage behaviour.

* Now require snakeoil version 0.2 and up- require new capability of 
  AtomicWriteFile, ability to specify uid/gid/perms.  Via that, fixes ticket
  109 (umask leaking through to profile.env).

* the 'glsa' pkgset is now deprecated in favor of 'vuln'; will remain
  through till 0.4 (ticket #106).
 
* ticket 105/96; fix via andkit, basically a bug in einstall lead to
  extra einstall opts getting dropped instead of passed through.

* compatibility fix for lha unpacking for nwere versions of lha.

* emake now invokes ${MAKE:-make}, instead of make- undocumented ebuild
  req, see bug 186598 at bugs.gentoo.org.

* pmerge --verbose is now pmerge -F portage-verbose-formatter

* Stop installing pregen symlink; functionality moved to pmaint regen.

* 'pmerge --domain' was added; basically is a way to specify the domain to
  use, else usees the configuration defined default domain.

* new ebuild trigger to avoid installing files into symlinked dir (get_libdir
  is the friend to fix a common /usr/lib -> /usr/lib64 bug), ticket 119


pkgcore 0.3.1:

* ticket 86; export FILE for portage_conf FETCHCOMMAND/RESUMECOMMAND support,
  convert from spawn_bash to spawn, add some extra error detection

* Correct cleanup of unknown state ebp processors; basically discard them if
  they fail in any way.  Cleanup inherit error msg when under ebd.

* Correct permission issue for vdb virtuals cache.

* ticket 84; rework overlay internals so that sorting order can't accidentally
  expose a version masked by a higher priority repository in an overlay stack.


pkgcore 0.3:

* pregen has moved into pmaint regen.


pkgcore/scripts/pconfig.py
# Copyright: 2006 Marien Zwart <marienz@gentoo.org>
# License: GPL2

"""Configuration querying utility."""


import traceback

from pkgcore.config import errors, basics
from pkgcore.plugin import get_plugins
from pkgcore.util import commandline
from snakeoil import modules

class DescribeClassParser(commandline.OptionParser):

    """Our option parser."""

    def check_values(self, values, args):
        values, args = commandline.OptionParser.check_values(
            self, values, args)
        if len(args) != 1:
            self.error('need exactly one argument: class to describe.')
        try:
            values.describe_class = modules.load_attribute(args[0])
        except modules.FailedImport, e:
            self.error(str(e))
        return values, ()


def dump_section(config, out):
    out.first_prefix.append('    ')
    out.write('# typename of this section: %s' % (config.type.name,))
    out.write('class %s.%s;' % (config.type.callable.__module__,
                                config.type.callable.__name__))
    if config.default:
        out.write('default true;')
    for key, val in sorted(config.config.iteritems()):
        typename = config.type.types.get(key)
        if typename is None:
            if config.type.allow_unknowns:
                typename = 'str'
            else:
                raise ValueError('no type set for %s (%r)' % (key, val))
        out.write('# type: %s' % (typename,))
        if typename.startswith('lazy_refs'):
            typename = typename[5:]
            val = list(ref.collapse() for ref in val)
        elif typename.startswith('lazy_ref'):
            typename = typename[5:]
            val = val.collapse()
        if typename == 'str':
            out.write('%s %r;' % (key, val))
        elif typename == 'bool':
            out.write('%s %s;' % (key, bool(val)))
        elif typename == 'list':
            out.write('%s %s;' % (
                    key, ' '.join(repr(string) for string in val)))
        elif typename == 'callable':
            out.write('%s %s.%s;' % (key, val.__module__, val.__name__))
        elif typename.startswith('ref:'):
            if val.name is None:
                out.write('%s {' % (key,))
                dump_section(val, out)
                out.write('};')
            else:
                out.write('%s %r;' % (key, val.name))
        elif typename.startswith('refs:'):
            out.autoline = False
            out.write('%s' % (key,))
            for i, subconf in enumerate(val):
                if subconf.name is None:
                    out.autoline = True
                    out.write(' {')
                    dump_section(subconf, out)
                    out.autoline = False
                    out.write('}')
                else:
                    out.write(' %r' % (subconf.name,))
            out.autoline = True
            out.write(';')
        else:
            out.write('# %s = %r of unknown type %s' % (key, val, typename))
    out.first_prefix.pop()


def get_classes(configs):
    # Not particularly efficient (doesn't memoize already visited configs)
    classes = set()
    for config in configs:
        classes.add('%s.%s' % (config.type.callable.__module__,
                               config.type.callable.__name__))
        for key, val in config.config.iteritems():
            typename = config.type.types.get(key)
            if typename is None:
                continue
            if typename.startswith('ref:'):
                classes.update(get_classes((val,)))
            elif typename.startswith('refs:'):
                classes.update(get_classes(val))
            elif typename.startswith('lazy_refs'):
                classes.update(get_classes(c.collapse() for c in val))
            elif typename.startswith('lazy_ref'):
                classes.update(get_classes((val.collapse(),)))
    return classes

def classes_main(options, out, err):
    """List all classes referenced by the config."""
    configmanager = options.config
    sections = []
    for name in configmanager.sections():
        try:
            sections.append(configmanager.collapse_named_section(name))
        except errors.CollapseInheritOnly:
            pass
    for classname in sorted(get_classes(sections)):
        out.write(classname)


def write_type(out, type_obj):
    out.write('typename is %s' % (type_obj.name,))
    if type_obj.doc:
        for line in type_obj.doc.split('\n'):
            out.write(line.strip(), wrap=True)
    if type_obj.allow_unknowns:
        out.write('values not listed are handled as strings')
    out.write()
    for name, typename in sorted(type_obj.types.iteritems()):
        out.write('%s: %s' % (name, typename), autoline=False)
        if name in type_obj.required:
            out.write(' (required)', autoline=False)
        out.write()


def describe_class_main(options, out, err):
    """Describe the arguments a class needs."""
    try:
        type_obj = basics.ConfigType(options.describe_class)
    except errors.TypeDefinitionError:
        err.write('Not a valid type!')
        return 1
    write_type(out, type_obj)


def uncollapsable_main(options, out, err):
    """Show things that could not be collapsed."""
    config = options.config
    for name in config.sections():
        try:
            config.collapse_named_section(name)
        except errors.CollapseInheritOnly:
            pass
        except errors.ConfigurationError, e:
            if options.debug:
                traceback.print_exc()
            else:
                out.write(str(e))
            out.write()


class _TypeNameParser(commandline.OptionParser):

    """Base for subcommands that take an optional type name."""

    def check_values(self, values, args):
        values, args = commandline.OptionParser.check_values(self, values,
                                                             args)
        if len(args) > 1:
            self.error('pass at most one typename')
        if args:
            values.typename = args[0]
        else:
            values.typename = None
        return values, ()


class DumpParser(_TypeNameParser):

    def __init__(self, **kwargs):
        # Make sure we do not pass two description kwargs if kwargs has one.
        kwargs['description'] = (
            'Dump the entire configuration. '
            'The format used is similar to the ini-like default '
            'format, but do not rely on this to always write a '
            'loadable config. There may be quoting issues. '
            'With a typename argument only that type is dumped.')
        kwargs['usage'] = '%prog [options] [typename]'
        _TypeNameParser.__init__(self, **kwargs)


def dump_main(options, out, err):
    """Dump the entire configuration."""
    config = options.config
    if options.typename is None:
        names = config.sections()
    else:
        names = getattr(config, options.typename).iterkeys()
    for name in sorted(names):
        try:
            section = config.collapse_named_section(name)
        except errors.CollapseInheritOnly:
            continue
        out.write('%r {' % (name,))
        dump_section(section, out)
        out.write('}')
        out.write()


class ConfigurablesParser(_TypeNameParser):

    def __init__(self, **kwargs):
        # Make sure we do not pass two description kwargs if kwargs has one.
        kwargs['description'] = (
            'List registered configurables (may not be complete). '
            'With a typename argument only configurables of that type are '
            'listed.')
        kwargs['usage'] = '%prog [options] [typename]'
        _TypeNameParser.__init__(self, **kwargs)


def configurables_main(options, out, err):
    """List registered configurables."""
    for configurable in get_plugins('configurable'):
        type_obj = basics.ConfigType(configurable)
        if options.typename is not None and type_obj.name != options.typename:
            continue
        out.write(out.bold, '%s.%s' % (
                configurable.__module__, configurable.__name__))
        write_type(out, type_obj)
        out.write()
        out.write()


def _dump_uncollapsed_section(config, out, err, section):
    """Write a single section."""
    if isinstance(section, basestring):
        out.write('named section %r' % (section,))
        return
    for key in sorted(section.keys()):
        kind, value = section.get_value(config, key, 'repr')
        out.write('# type: %s' % (kind,))
        if kind == 'list':
            for name, val in zip((
                    key + '.prepend', key, key + '.append'), value):
                if val:
                    out.write(
                        repr(name), ' = ', ' '.join(repr(v) for v in val))
            continue
        if kind in ('refs', 'str'):
            for name, val in zip((
                    key + '.prepend', key, key + '.append'), value):
                if not val:
                    continue
                out.write(repr(name), ' = ', autoline=False)
                if kind == 'str':
                    out.write(repr(val))
                else:
                    out.write()
                    out.first_prefix.append('    ')
                    try:
                        for subnr, subsection in enumerate(val):
                            subname = 'nested section %s' % (subnr + 1,)
                            out.write(subname)
                            out.write('=' * len(subname))
                            _dump_uncollapsed_section(config, out, err, subsection)
                            out.write()
                    finally:
                        out.first_prefix.pop()
            continue
        out.write('%r = ' % (key,), autoline=False)
        if kind == 'callable':
            out.write(value.__module__, value.__name__)
        elif kind == 'bool':
            out.write(str(value))
        elif kind == 'ref':
            out.first_prefix.append('    ')
            try:
                out.write()
                _dump_uncollapsed_section(config, out, err, value)
            finally:
                out.first_prefix.pop()
        else:
            err.error('unsupported type %r' % (kind,))


def dump_uncollapsed_main(options, out, err):
    """dump the configuration in a raw, uncollapsed form.
    Not directly usable as a configuration file, mainly used for inspection
    """
    out.write('''# Warning:
# Do not copy this output to a configuration file directly,
# because the types you see here are only guesses.
# A value used as "list" in the collapsed config will often
# show up as "string" here and may need to be converted
# (for example from space-separated to comma-separated)
# to work in a config file with a different format.
''')
    for i, source in enumerate(options.config.configs):
        s = 'Source %s' % (i + 1,)
        out.write(out.bold, '*' * len(s))
        out.write(out.bold, s)
        out.write(out.bold, '*' * len(s))
        out.write()
        for name, section in sorted(source.iteritems()):
            out.write('%s' % (name,))
            out.write('=' * len(name))
            _dump_uncollapsed_section(options.config, out, err, section)
            out.write()


commandline_commands = {
    'dump': (DumpParser, dump_main),
    'classes': (commandline.OptionParser, classes_main),
    'uncollapsable': (commandline.OptionParser, uncollapsable_main),
    'describe_class': (DescribeClassParser, describe_class_main),
    'configurables': (ConfigurablesParser, configurables_main),
    'dump-uncollapsed': (commandline.OptionParser, dump_uncollapsed_main),
    }
